api/v2/routes_payment.go

```diff
@@ -290,6 +293,57 @@ func (api *API) CreateDashPayment(c *gin.Context) {
 	Respond(c, http.StatusOK, gin.H{"response": p})
 }
 
+func (api *API) stripeCharge(c *gin.Context) {
+	username, err := GetAuthenticatedUserFromContext(c)
+	if err != nil {
+		api.LogError(c, err, eh.NoAPITokenError)(http.StatusBadRequest)
+		return
+	}
+	forms, missingField := api.extractPostForms(c, "stripe_token", "stripe_email", "value_in_cents")
+	if missingField != "" {
+		FailWithMissingField(c, missingField)
+		return
+	}
+	valueInCentsInt, err := strconv.ParseInt(forms["value_in_cents"], 10, 64)
+	if err != nil {
+		Fail(c, err)
+		return
+	}
+	valueInCentsFloat, err := strconv.ParseFloat(forms["value_in_cents"], 64)
+	if err != nil {
+		Fail(c, err)
+	}
+	// set the secret key after validation
+	stripe.Key = api.cfg.Stripe.SecretKey
+	// initialize credit card charge parameters
+	chargeParams := &stripe.ChargeParams{
+		Amount:      stripe.Int64(valueInCentsInt),
+		Currency:    stripe.String(string(stripe.CurrencyUSD)),
+		Description: stripe.String("temporal credit purchase"),
+		// StatementDescriptor is what appears in their credit card billing report
+		StatementDescriptor: stripe.String("credit purchase"),
+		// email the receipt goes to
+		ReceiptEmail: stripe.String(forms["stripe_email"]),
+	}
+	// set the token generated by stripe
+	chargeParams.SetSource(forms["stripe_token"])
+	// metadata can be used for better Radar validation
+	chargeParams.AddMetadata("order_type", "temporal.credits")
```

Comment:
can't these be set directly when we make `ChargeParams`?

Reply:
no they have they have to be configured after instantiating `ChargeParams`

Reply:
looks like you can (and this is safer too, protecting against invalid source):

```go
	// set the token generated by stripe
	source, err := stripe.SourceParamsFor(forms["stripe_token"])
	if err != nil {
		Fail(c, err)
		return
	}

	// initialize credit card charge parameters
	chargeParams := &stripe.ChargeParams{
		Amount:      stripe.Int64(valueInCentsInt),
		Currency:    stripe.String(string(stripe.CurrencyUSD)),
		Description: stripe.String("temporal credit purchase"),
		// StatementDescriptor is what appears in their credit card billing report
		StatementDescriptor: stripe.String("credit purchase"),
		// email the receipt goes to
		ReceiptEmail: stripe.String(forms["stripe_email"]),
		Source:       source,
		Params: stripe.Params{
			// metadata can be used for better Radar validation
			Metadata: map[string]string{"order_type": "temporal.credits"},
		},
	}
```

Reply:
this gets rid of the setters, and you can now line the whole thing to make it easier to read:

```go
	// set the token generated by stripe
	source, err := stripe.SourceParamsFor(forms["stripe_token"])
	if err != nil {
		Fail(c, err)
		return
	}

	// create charge
	ch, err := charge.New(&stripe.ChargeParams{
		Amount:      stripe.Int64(valueInCentsInt),
		Currency:    stripe.String(string(stripe.CurrencyUSD)),
		Description: stripe.String("temporal credit purchase"),
		// StatementDescriptor is what appears in their credit card billing report
		StatementDescriptor: stripe.String("credit purchase"),
		// email the receipt goes to
		ReceiptEmail: stripe.String(forms["stripe_email"]),
		Source:       source,
		Params: stripe.Params{
			// metadata can be used for better Radar validation
			Metadata: map[string]string{"order_type": "temporal.credits"},
		},
	})
```

Reply:
source: https://sourcegraph.com/github.com/RTradeLtd/Temporal@payments/stripe/-/blob/vendor/github.com/stripe/stripe-go/charge.go#L86:28

Reply:
slick